packages/core/utils/src/validate/validators.ts

```diff
@@ -0,0 +1,112 @@
+import { curry, isEmpty, isNil, isObject } from 'lodash/fp';
+
+import { pipeAsync } from '../async';
+import traverseEntity, { Data } from '../traverse-entity';
+import { isScalarAttribute } from '../content-types';
+
+import traversals from '../traverse/traversals';
+
+import { throwPassword, throwPrivate, throwDynamicZones, throwMorphToRelations } from './visitors';
+import { isOperator } from '../operators';
+
+import type { Model } from '../types';
+import { ValidationError } from '../errors';
+
+const { traverseQueryFilters, traverseQuerySort, traverseQueryFields } = traversals;
+
+const throwPasswords = (schema: Model) => async (entity: Data) => {
+  return traverseEntity(throwPassword, { schema }, entity);
+};
+
+const defaultValidateFilters = curry((schema: Model, filters: unknown) => {
+  return pipeAsync(
+    // Remove keys that are not attributes or valid operators
+    traverseQueryFilters(
+      ({ key, attribute }) => {
+        const isAttribute = !!attribute;
+
+        if (!isAttribute && !isOperator(key) && key !== 'id') {
+          throw new ValidationError(`invalid key ${key}`);
+        }
+      },
+      { schema }
+    ),
+    // Remove dynamic zones from filters
+    traverseQueryFilters(throwDynamicZones, { schema }),
+    // Remove morpTo relations from filters
+    traverseQueryFilters(throwMorphToRelations, { schema }),
+    // Remove passwords from filters
+    traverseQueryFilters(throwPassword, { schema }),
+    // Remove private from filters
+    traverseQueryFilters(throwPrivate, { schema }),
+    // Remove empty objects
+    traverseQueryFilters(
+      ({ key, value }) => {
+        if (isObject(value) && isEmpty(value)) {
+          throw new ValidationError(`invalid key ${key}`);
+        }
+      },
+      { schema }
```

Comment:
Maybe you can create an array and map it with the errors like (I don't know what looks cleaner)
`[throwPrivate, throwPassword, ... ].map(throwError => traverseQueryFilters(throwError, {schema})`